File: renewals-never-run/python/run_due_renewals.py
"""Trigger WooCommerce Subscriptions renewals that Action Scheduler stopped running.

When the Action Scheduler queue stalls (a fatal error in one action, a maxed out
worker, a cron that stopped firing) the scheduled-subscription-payment actions pile
up "pending" long past their scheduled_date. WooCommerce never asked Stripe for the
money, so the subscription just sits there looking active while nothing is billed.

This walks orders that look like stuck renewals, reads the saved Stripe PaymentIntent
(or the customer's saved payment method) and charges the renewal amount directly
through the Stripe API, then reports the result back onto the order. Read the queue
through Action Scheduler's own table via a small custom endpoint, or, if you do not
want a custom endpoint, treat any subscription whose "next payment" date is in the
past as due. Safe to run again and again. Read only in DRY_RUN mode.
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("run_due_renewals")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
GRACE_HOURS = int(os.environ.get("GRACE_HOURS", "3"))
STALE_DAYS = int(os.environ.get("STALE_DAYS", "14"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

RENEWABLE_STATUSES = {"active", "on-hold"}


def intent_id_of(order):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid if tid and tid.startswith("pi_") else None


def order_amount_minor(order):
    # Works for two decimal currencies. Zero decimal currencies (JPY and friends)
    # have their own guide, since 50.00 is wrong for those.
    return round(float(order["total"]) * 100)


def hours_overdue(scheduled_ts, now_ts):
    return (now_ts - scheduled_ts) / 3600


def decide(subscription, now_ts, grace_hours=GRACE_HOURS, stale_days=STALE_DAYS):
    """Pure decision: what to do about one subscription's due renewal.

    subscription is a plain dict with:
      status: the subscription status string
      next_payment_ts: unix timestamp the renewal was scheduled for, or None
      last_order_status: status of the most recent renewal order, or None
      payment_method_token: a saved Stripe payment method id, or None

    Returns (action, reason) where action is one of:
      "skip"    - nothing due, or already handled
      "wait"    - due, but still inside the grace window, leave it to the scheduler
      "charge"  - due, past grace, and we have what we need to charge it
      "blocked" - due, past grace, but there is no saved payment method to charge
      "stale"   - overdue so long it needs a human, not an auto charge
    """
    if subscription["status"] not in RENEWABLE_STATUSES:
        return ("skip", "subscription is not active or on-hold")
    if subscription.get("next_payment_ts") is None:
        return ("skip", "no renewal scheduled")
    if subscription["last_order_status"] in ("processing", "completed"):
        return ("skip", "renewal already paid")

    overdue_hours = hours_overdue(subscription["next_payment_ts"], now_ts)
    if overdue_hours < 0:
        return ("skip", "renewal is not due yet")
    if overdue_hours < grace_hours:
        return ("wait", "inside the grace window, scheduler may still catch it")
    if overdue_hours >= stale_days * 24:
        return ("stale", "overdue longer than the stale window, needs a human look")
    if not subscription.get("payment_method_token"):
        return ("blocked", "no saved payment method to charge")
    return ("charge", "past due and past grace, safe to charge now")


def get_subscription(subscription_id):
    r = requests.get(
        f"{WOO_URL}/wp-json/wc/v3/subscriptions/{subscription_id}",
        auth=AUTH, timeout=30,
    )
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def due_subscriptions():
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/subscriptions",
            params={"status": "active,on-hold", "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for sub in batch:
            yield sub
        page += 1


def charge_renewal(subscription, order):
    """Charge the saved payment method off-session for the renewal amount."""
    amount = order_amount_minor(order)
    intent = stripe.PaymentIntent.create(
        amount=amount,
        currency=order.get("currency", "usd").lower(),
        customer=subscription["customer_stripe_id"],
        payment_method=subscription["payment_method_token"],
        off_session=True,
        confirm=True,
        metadata={"order_id": str(order["id"]), "subscription_id": str(subscription["id"])},
    )
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}",
        json={"status": "processing", "transaction_id": intent.id},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={"note": f"Renewal charged manually after Action Scheduler stalled. "
                      f"Stripe PaymentIntent {intent.id}, status {intent.status}."},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    return intent


def run():
    charged = 0
    now_ts = int(time.time())
    for sub in due_subscriptions():
        order = sub.get("last_order") or {}
        record = {
            "status": sub["status"],
            "next_payment_ts": sub.get("next_payment_ts"),
            "last_order_status": order.get("status"),
            "payment_method_token": sub.get("payment_method_token"),
        }
        action, reason = decide(record, now_ts)
        if action in ("skip", "wait"):
            continue
        if action in ("blocked", "stale"):
            log.warning("Subscription %s: %s. %s", sub["id"], action, reason)
            continue
        log.info("Subscription %s: %s. %s", sub["id"], reason, "would charge" if DRY_RUN else "charging")
        if not DRY_RUN:
            charge_renewal(sub, order)
        charged += 1
    log.info("Done. %d renewal(s) %s.", charged, "to charge" if DRY_RUN else "charged")


if __name__ == "__main__":
    run()

File: renewals-never-run/node/run-due-renewals.js
/**
 * Trigger WooCommerce Subscriptions renewals that Action Scheduler stopped running.
 *
 * When the Action Scheduler queue stalls (a fatal error in one action, a maxed out
 * worker, a cron that stopped firing) the scheduled-subscription-payment actions
 * pile up "pending" long past their scheduled_date. WooCommerce never asked Stripe
 * for the money, so the subscription just sits there looking active while nothing
 * is billed.
 *
 * This walks subscriptions that look like stuck renewals, reads the saved Stripe
 * payment method, and charges the renewal amount directly through the Stripe API,
 * then reports the result back onto the order. Safe to run again and again.
 * Read only in DRY_RUN mode.
 *
 * Guide: https://www.allanninal.dev/woocommerce/renewals-never-run/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const GRACE_HOURS = Number(process.env.GRACE_HOURS || 3);
const STALE_DAYS = Number(process.env.STALE_DAYS || 14);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const RENEWABLE_STATUSES = new Set(["active", "on-hold"]);

export function intentIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = order.transaction_id;
  return tid && tid.startsWith("pi_") ? tid : null;
}

export function orderAmountMinor(order) {
  // Works for two decimal currencies. Zero decimal currencies (JPY and friends)
  // have their own guide, since 50.00 is wrong for those.
  return Math.round(parseFloat(order.total) * 100);
}

function hoursOverdue(scheduledTs, nowTs) {
  return (nowTs - scheduledTs) / 3600;
}

/**
 * Pure decision: what to do about one subscription's due renewal.
 *
 * subscription is a plain object with:
 *   status: the subscription status string
 *   nextPaymentTs: unix timestamp (seconds) the renewal was scheduled for, or null
 *   lastOrderStatus: status of the most recent renewal order, or null
 *   paymentMethodToken: a saved Stripe payment method id, or null
 *
 * Returns [action, reason] where action is one of:
 *   "skip"    - nothing due, or already handled
 *   "wait"    - due, but still inside the grace window, leave it to the scheduler
 *   "charge"  - due, past grace, and we have what we need to charge it
 *   "blocked" - due, past grace, but there is no saved payment method to charge
 *   "stale"   - overdue so long it needs a human, not an auto charge
 */
export function decide(subscription, nowTs, graceHours = GRACE_HOURS, staleDays = STALE_DAYS) {
  if (!RENEWABLE_STATUSES.has(subscription.status)) {
    return ["skip", "subscription is not active or on-hold"];
  }
  if (subscription.nextPaymentTs === null || subscription.nextPaymentTs === undefined) {
    return ["skip", "no renewal scheduled"];
  }
  if (["processing", "completed"].includes(subscription.lastOrderStatus)) {
    return ["skip", "renewal already paid"];
  }

  const overdueHours = hoursOverdue(subscription.nextPaymentTs, nowTs);
  if (overdueHours < 0) return ["skip", "renewal is not due yet"];
  if (overdueHours < graceHours) return ["wait", "inside the grace window, scheduler may still catch it"];
  if (overdueHours >= staleDays * 24) return ["stale", "overdue longer than the stale window, needs a human look"];
  if (!subscription.paymentMethodToken) return ["blocked", "no saved payment method to charge"];
  return ["charge", "past due and past grace, safe to charge now"];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* dueSubscriptions() {
  let page = 1;
  while (true) {
    const batch = await woo(`/subscriptions?status=active,on-hold&per_page=50&page=${page}`);
    if (!batch || !batch.length) return;
    for (const sub of batch) yield sub;
    page++;
  }
}

async function chargeRenewal(subscription, order) {
  const amount = orderAmountMinor(order);
  const intent = await stripe.paymentIntents.create({
    amount,
    currency: (order.currency || "usd").toLowerCase(),
    customer: subscription.customer_stripe_id,
    payment_method: subscription.payment_method_token,
    off_session: true,
    confirm: true,
    metadata: { order_id: String(order.id), subscription_id: String(subscription.id) },
  });
  await woo(`/orders/${order.id}`, {
    method: "PUT",
    body: JSON.stringify({ status: "processing", transaction_id: intent.id }),
  });
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Renewal charged manually after Action Scheduler stalled. ` +
            `Stripe PaymentIntent ${intent.id}, status ${intent.status}.`,
    }),
  });
  return intent;
}

export async function run() {
  let charged = 0;
  const nowTs = Math.floor(Date.now() / 1000);
  for await (const sub of dueSubscriptions()) {
    const order = sub.last_order || {};
    const record = {
      status: sub.status,
      nextPaymentTs: sub.next_payment_ts ?? null,
      lastOrderStatus: order.status ?? null,
      paymentMethodToken: sub.payment_method_token ?? null,
    };
    const [action, reason] = decide(record, nowTs);
    if (action === "skip" || action === "wait") continue;
    if (action === "blocked" || action === "stale") {
      console.warn(`Subscription ${sub.id}: ${action}. ${reason}`);
      continue;
    }
    console.log(`Subscription ${sub.id}: ${reason}. ${DRY_RUN ? "would charge" : "charging"}`);
    if (!DRY_RUN) await chargeRenewal(sub, order);
    charged++;
  }
  console.log(`Done. ${charged} renewal(s) ${DRY_RUN ? "to charge" : "charged"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}
